controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,349 @@
+	if len(r.apiResourceList) == 0 ||
+		r.discoveryLastRefreshed.Add(time.Minute*10).Before(time.Now()) {
+		err := r.refreshDiscoveryInfo()
+		if err != nil {
+			return "", err
+		}
+
+		r.discoveryLastRefreshed = time.Now()
+
+		// The discovery is just refeshed so skip another refesh
+		skipRefresh = true
+	}
+
+	for _, resc := range r.apiResourceList {
+		groupVerison, err := schema.ParseGroupVersion(resc.GroupVersion)
+		if err != nil {
+			r.Log.Error(err, "Cannot parse the group and version in getApiVersion ", "GroupVersion:", resc.GroupVersion)
+
+			continue
+		}
+
+		group := groupVerison.Group
+		version := groupVerison.Version
+		// Consider groupversion == v1 or groupversion == app1/v1
+		for _, apiResource := range resc.APIResources {
+			if apiResource.Kind == kind && group == apiGroup {
+				return version, nil
+			}
+		}
+	}
+
+	if !skipRefresh {
+		// Get new discoveryInfo, when any resource is not found
+		err := r.refreshDiscoveryInfo()
+		if err != nil {
+			return "", err
+		}
+
+		r.discoveryLastRefreshed = time.Now()
+		// Retry one more time after refresh the discovery
+		return r.getAPIVersion(kind, apiGroup, true)
+	}
+
+	return "", ErrNotFoundDiscovery
+}
+
+// Retrieve all groups and versions to add in config sync
+// Constraints present only kind and group so this function helps to find the version
+func (r *ConstraintStatusReconciler) refreshDiscoveryInfo() error {
+	discoveryClient := r.ClientSet.Discovery()
+
+	apiList, err := discoveryClient.ServerPreferredResources()
+	if err != nil {
+		return err
+	}
+
+	// Save fetched discovery at apiResourceList
+	r.apiResourceList = apiList
+
+	return nil
+}
+
+func (r *ConstraintStatusReconciler) getUniqSyncOnly() []v1alpha1.SyncOnlyEntry {
+	syncOnlySet := map[v1alpha1.SyncOnlyEntry]bool{}
+	// Add to table for unique filtering
+	for _, syncEntries := range r.ConstraintToSyncOnly {
+		for _, entry := range syncEntries {
+			syncOnlySet[entry] = true
+		}
+	}
+
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+	for key := range syncOnlySet {
+		syncOnlys = append(syncOnlys, key)
+	}
+
+	return syncOnlys
+}
+
+func (r *ConstraintStatusReconciler) updateConfigSyncOnlyEntry(
+	ctx context.Context, config *v1alpha1.Config, syncOnlys []v1alpha1.SyncOnlyEntry,
+) error {
+	config.Spec.Sync.SyncOnly = syncOnlys
+
+	return r.Update(ctx, config, &client.UpdateOptions{})
+}
+
+func (r *ConstraintStatusReconciler) addConstraintPodStatusToMap(cpsName string, entries []v1alpha1.SyncOnlyEntry) {
+	r.ConstraintToSyncOnly[cpsName] = entries
+}
+
+// This function is called when constraintPodStatus deleted event.
+// Delete constraintPodstatus key and refresh r.constraintMap with current constraint
+func (r *ConstraintStatusReconciler) handleDeleteEvent(
+	ctx context.Context, cpsName string, config *v1alpha1.Config,
+) error {
+	delete(r.ConstraintToSyncOnly, cpsName)
+
+	updatedSyncOnly := r.getUniqSyncOnly()
+
+	err := r.updateConfigSyncOnlyEntry(ctx, config, updatedSyncOnly)
```

Comment:
You should only update if the syncOnly changed. Perhaps another constraint has the same match.

Reply:
Duplicated removed at getUniqSyncOnly

Reply:
I have an e2e test for this